test/release/examples/primers/modules.chpl

```diff
@@ -0,0 +1,390 @@
+/*
+ * Module Primer
+ *
+ * This primer introduces the concept of modules, a strategy for encapsulating
+ * code for later use.  It covers:
+ *   how to define a module
+ *   namespace control for a module's symbols
+ *   outside access of a module's symbols
+ *   namespace control when using a module in an outside context, including:
+ *     unlimited
+ *     explicit exclusion of symbols
+ *     explicit inclusion of symbols
+ *     renaming included symbols
+ *   outside access of modules in different files
+ */
+
+
+/* A module is a grouping of code - each program consists of at least one
+   module, and every symbol is associated with some module.  If all the code of
+   a file is not enclosed in an explicit module, defined using the syntax below,
+   then the file itself is treated as a module with the same name as the file
+   (minus the .chpl suffix).
+ */
+module modToUse {
+
+  /* In this case, foo is a public global variable which lives in the module
+     modToUse */
+  var foo = 12;
+
+  // As is bar.
+  var bar: int = 2;
+
+  /* baz is a public function which lives in modToUse */
+  proc baz (x, y) {
+    return x * (x + y);
+  }
+
+  /* A module defined within another module is known as a nested module.  These
+     modules can take advantage of symbols defined within their parent module,
+     but their parent module must take additional steps to access the contents
+     of the nested module.  We will cover how to access the contents of other
+     modules later in this primer.
+  */
+  module Inner1 {
+    /* The variable foobar is utilizing modToUse's foo and bar variables. */
+    var foobar = foo + bar;
+  }
+
+  /* Note that since there is other code after the end of modToUse's definition,
+     modToUse is a submodule of the overarching file module "modules".
```

Comment:
Oh, I remember the case I was thinking of...  This program generates a warning because a lot of users equate 'use' with '#include' and so think it's OK to put 'use's anywhere:

    use Math;
    module MyProgram {
      writeln("Hello, world!");
    }

Then they got surprised when the code above resulted in a no-op execution (because there was an empty outer module and nothing ever use'd / initialized the inner module).

I think I would probably take the approach of wrapping all the code that exercises the modules that you set up in a sibling module of its own (named something indicative of its role as the driver or the main module or something...) so that the file only contains module definitions.  That avoids needing to talk about the impact of the file-scoped code and the implicit module here (and I think the earlier description you had of filenames being treated as module names if there is no top-level code covers that topic well enough).